### DISCUSSION TITLE
Usage as Library to Generate a PDF from scratch ...

### DISCUSSION THREAD

**Discussion by a2800276** [pdfcpu/pdfcpu#570] [General]
I'm banging my head trying to create a simple PDF from: a.)  go code and b.) scratch

What I mean is just single A4 page  that says "Hello World" on it.  

I feels like I've overlooked some obvious bit of documentation that describes this ...
or maybe that this is just not the intended usage of pdf cpu, and it's meant to modify existing documents? Everything is so wonderfully documented for the CLI case that I was hoping the PDF-Fairy would help me avoid any hard work...

The best I've code example of creating pdfs I've been able to find are in the `api/test/create_test.go`

And the process would roughly look like this:

- create an A4  `types.Rectangle`
- ... to be used as MediaBox of `model.Page`
- ...create and XrefTable and extract it's root
- and now my head hurts trying to figure out how the TextDesciptor I require to add my Hello World string ties into everything :)

Is there a simpler "Getting Started" example that I'm missing? Or a "friendlier" interface (the `pdf/primitives/ directory looks promising) ...  If not, would it make sense to try to contribute documenation, or is PDF creation just not the intented focus of the library?  And could someone point me in the correct direction of how the library is intended to be used for document creation?

**Comment by hhrutter** [comment]
Hi there,

The preferred way to create PDFs within pdfcpu is using JSON.

There is no official api for low level PDF creation and that's why there is no "Getting started for PDF creation" but that does not mean the code is not there. 

You are on the right track though and these should help you to get going:

`pkg/api/test/create_test.go`

`pkg/samples/basic/*`

**Reply by a2800276** [reply]
Puh.Glad I did not make a moronic mistake. Thanks for the quick response and the excellent work in general! This is a very impressive project!

**Comment by mkungla** [comment]
@hhrutter wierd looking into library I ended up here in this topic before finding refreence to JSON schema to compose pdf documents

> The preferred way to create PDFs within pdfcpu is using JSON.

I'm experimenting to build [Gio UI](https://gioui.org/) based pdf document composer from templates using this libray under the hood, but after few hours going trough documentation and code I am still confused what would be right way to approach it.

**Comment by hhrutter** [comment]
Hi, that's because pdfcpu is foremost a CLI tool and only getting ready for smth which can be called Beta.
The API needs substantial finishing and the docs for PDF creation are coming along slower than I'd hoped for due to lack of time on my part. Meanwhile if you state a specific question maybe I can help.

**Comment by hhrutter** [comment]
Closed due to inactivity
